File: mindspore/python/mindspore/ops/_grad_experimental/grad_image_ops.py
# Copyright 2021-2022 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

"""image_ops"""

from mindspore import Tensor
from mindspore.common import dtype as mstype
from mindspore.ops._grad.grad_base import bprop_getters
from mindspore.ops import operations as P
from mindspore.ops import functional as F
from mindspore.ops.operations import _grad_ops as G
from mindspore.ops.operations.image_ops import ResizeBicubic
from mindspore.ops.operations._grad_ops import ResizeBicubicGrad
from mindspore.ops.operations.image_ops import ResizeV2
from mindspore.ops.composite.multitype_ops.zeros_like_impl import zeros_like
from mindspore.ops.operations.image_ops import CropAndResize
from mindspore.ops.operations.image_ops import CropAndResizeGradImage
from mindspore.ops.operations.image_ops import CropAndResizeGradBoxes
from mindspore.ops.operations.image_ops import RGBToHSV
from mindspore.ops.operations.image_ops import ScaleAndTranslate
from mindspore import context


@bprop_getters.register(ResizeBicubic)
def get_bprop_resize_bicubic(self):
    """Grad definition for `ResizeBicubic` operation."""
    resize_bicubic_grad = ResizeBicubicGrad(align_corners=self.align_corners,
                                            half_pixel_centers=self.half_pixel_centers)

    def bprop(images, size, out, dout):
        images_type = F.dtype(images)
        type_list = [mstype.int8, mstype.uint8, mstype.int16, mstype.uint16, mstype.int32,
                     mstype.int64, mstype.float16]
        if images_type in type_list:
            images = F.cast(images, mstype.float64)
        dx = resize_bicubic_grad(dout, images)
        return (dx, P.ZerosLike()(size))
    return bprop


@bprop_getters.register(ResizeV2)
def get_bprop_resize_v2(self):
    """Grad definition for `ResizeV2` operation."""
    resize_v2_grad = G.ResizeV2Grad(coordinate_transformation_mode=self.coordinate_transformation_mode,
                                    mode=self.mode)

    def bprop(x, roi, scales, sizes, out, dout):
        input_size = P.Shape()(x)
        dx = resize_v2_grad(dout, roi, scales, Tensor(input_size))
        return (dx, zeros_like(roi), zeros_like(scales), zeros_like(sizes))
    return bprop


@bprop_getters.register(CropAndResize)
def get_bprop_crop_and_resize(self):
    """Grad definition for `CropAndResize` operation."""
    allowed_types = [mstype.float16, mstype.float32, mstype.float64]
    gradboxes = CropAndResizeGradBoxes(method="bilinear")
    method_ = self.method
    dyn_shape = P.TensorShape()

    is_ascend_cpu = context.get_context('device_target') in ("Ascend", "CPU")
    def bprop(x, boxes, box_index, crop_size, out, dout):
        if method_ != "bilinear":
            if not is_ascend_cpu:
                return (zeros_like(x), zeros_like(boxes), zeros_like(box_index), zeros_like(crop_size))
        image_type = x.dtype
        if image_type not in allowed_types:
            x = F.cast(x, mstype.float32)
        dimage_type = image_type
        gradimage = CropAndResizeGradImage(dimage_type, method=method_)
        image_shape = x.shape
        if F.is_sequence_value_unknown(image_shape):
            image_size = dyn_shape(x)
            image_size = F.cast(image_size, mstype.int32)
        else:
            image_size = Tensor(image_shape, dtype=mstype.int32)
        dimage = gradimage(dout, boxes, box_index, image_size)
        dbox = gradboxes(dout, x, boxes, box_index)
        return (dimage, dbox, zeros_like(box_index), zeros_like(crop_size))
    return bprop


def crcp(x):
    """Grad definition for `RGBToHSV` operations."""
    return P.DivNoNan()(1, x)


def function1_rgbtohsv(images, out, dout):
    """Grad definition for `RGBToHSV` operations."""
    dout = P.Cast()(dout, mstype.float32)
    images = P.Cast()(images, mstype.float32)
    out = P.Cast()(out, mstype.float32)
    return images, out, dout


def function2_rgbtohsv(images):
    """Grad definition for `RGBToHSV` operations."""
    # Input Channels
    reds = images[..., 0]
    greens = images[..., 1]
    blues = images[..., 2]
    return reds, greens, blues


def function3_rgbtohsv(out, reds):
    """Grad definition for `RGBToHSV` operations."""
    # Output Channels
    saturation = out[..., 1]
    value = out[..., 2]
    dsr1 = P.Cast()(reds > 0, mstype.float32)
    return dsr1, saturation, value


def function4_rgbtohsv(reds, greens, blues):
    """Grad definition for `RGBToHSV` operations."""
    r_b = P.LogicalAnd()((reds >= blues), (reds >= greens))
    red_biggest = P.Cast()(r_b, mstype.float32)
    g_b = P.LogicalAnd()((greens > reds), (greens >= blues))
    green_biggest = P.Cast()(g_b, mstype.float32)
    b_b = P.LogicalAnd()((blues > reds), (blues > greens))
    blue_biggest = P.Cast()(b_b, mstype.float32)
    return red_biggest, green_biggest, blue_biggest


def function5_rgbtohsv(reds, greens, blues):
    """Grad definition for `RGBToHSV` operations."""
    r_s = P.LogicalAnd()((reds < blues), (reds < greens))
    red_smallest = P.Cast()(r_s, mstype.float32)
    g_s = P.LogicalAnd()((greens <= reds), (greens < blues))
    green_smallest = P.Cast()(g_s, mstype.float32)
    b_s = P.LogicalAnd()((blues <= reds), (blues <= greens))
    blue_smallest = P.Cast()(b_s, mstype.float32)
    return red_smallest, green_smallest, blue_smallest


def function6_rgbtohsv(red_biggest, green_biggest, blue_biggest):
    """Grad definition for `RGBToHSV` operations."""
    dv_dr = red_biggest
    dv_dg = green_biggest
    dv_db = blue_biggest
    return dv_dr, dv_dg, dv_db


def function7_rgbtohsv(greens, green_biggest, dhb5, dh_db_1, dh_db_2, dh_db_3, dh_db_4,\
              dout, dv_dr, dv_dg, dv_db, ds_dr, ds_dg, ds_db, dh_dr, dh_dg):
    """Grad definition for `RGBToHSV` operations."""
    dh_db_5 = 60 * (P.Cast()((greens > 0), mstype.float32) * green_biggest * dhb5)

    dh_db = dh_db_1 + dh_db_2 + dh_db_3 + dh_db_4 + dh_db_5

    dh_db = dh_db / 360

    dv_drgb = P.Stack(-1)(
        [dout[..., 2] * dv_dr, dout[..., 2] * dv_dg, dout[..., 2] * dv_db])
    ds_drgb = P.Stack(-1)(
        [dout[..., 1] * ds_dr, dout[..., 1] * ds_dg, dout[..., 1] * ds_db])
    dh_drgb = P.Stack(-1)(
        [dout[..., 0] * dh_dr, dout[..., 0] * dh_dg, dout[..., 0] * dh_db])
    dvds_drgb = P.Add()(dv_drgb, ds_drgb)
    doutient_input = P.Add()(dvds_drgb, dh_drgb)
    return (doutient_input,)


@bprop_getters.register(RGBToHSV)
def get_bprop_rgb_to_hsv(self):
    """dout definition for 'RGBToHSV' operation"""

    def bprop(images, out, dout):
        images, out, dout = function1_rgbtohsv(images, out, dout)
        reds, greens, blues = function2_rgbtohsv(images)
        dsr1, saturation, value = function3_rgbtohsv(out, reds)
        red_biggest, green_biggest, blue_biggest = function4_rgbtohsv(reds, greens, blues)
        red_smallest, green_smallest, blue_smallest = function5_rgbtohsv(reds, greens, blues)
        dv_dr, dv_dg, dv_db = function6_rgbtohsv(red_biggest, green_biggest, blue_biggest)
        dsr2 = red_biggest * P.Add()(green_smallest * greens, blue_smallest * blues) * crcp(P.Square()(reds))
        dsr3 = red_smallest * -1 * crcp((green_biggest * greens) + (blue_biggest * blues))
        ds_dr = dsr1 * P.Add()(dsr2, dsr3)
        dsg1 = P.Cast()((greens > 0), mstype.float32)
        dsg2 = green_biggest * P.Add()(red_smallest * reds, blue_smallest * blues) * crcp(P.Square()(greens))
        dsg3 = green_smallest * -1 * crcp((red_biggest * reds) + (blue_biggest * blues))
        ds_dg = dsg1 * P.Add()(dsg2, dsg3)

        dsb1 = P.Cast()((blues > 0), mstype.float32)
        dsb2 = blue_biggest * P.Add()(green_smallest * greens, red_smallest * reds) * crcp(P.Square()(blues))
        dsb3 = blue_smallest * -1 * crcp((green_biggest * greens) + (red_biggest * reds))
        ds_db = dsb1 * P.Add()(dsb2, dsb3)

        dhr1 = (greens - blues) * crcp(P.Square()(saturation)) * crcp(P.Square()(value))
        dh_dr_1 = 60 * (P.Cast()((reds > 0), mstype.float32) * red_biggest * -1 * dhr1)
        dhr2 = red_smallest * (blues - greens) * crcp(P.Square()(reds - greens))
        dh_dr_2 = 60 * (P.Cast()((greens > 0), mstype.float32) * green_biggest * dhr2)
        dhr3 = blue_smallest * -1 * crcp(greens - blues)
        dh_dr_3 = 60 * (P.Cast()((greens > 0), mstype.float32) * green_biggest * dhr3)
        dhr4 = red_smallest * (blues - greens) * crcp(P.Square()(blues - reds))
        dh_dr_4 = 60 * (P.Cast()((blues > 0), mstype.float32) * blue_biggest * dhr4)
        dhr5 = green_smallest * crcp(blues - greens)
        dh_dr_5 = 60 * (P.Cast()((blues > 0), mstype.float32) * blue_biggest * dhr5)

        dh_dr = (dh_dr_1 + dh_dr_2 + dh_dr_3 + dh_dr_4 + dh_dr_5) / 360

        dhg1 = (blues - reds) * crcp(P.Square()(saturation)) * crcp(P.Square()(value))
        dh_dg_1 = 60 * (P.Cast()((greens > 0), mstype.float32) * green_biggest * -1 * dhg1)
        dhg2 = green_smallest * (reds - blues) * crcp(P.Square()(reds - greens))
        dh_dg_2 = 60 * (P.Cast()((reds > 0), mstype.float32) * red_biggest * dhg2)
        dhg3 = blue_smallest * crcp(reds - blues)
        dh_dg_3 = 60 * (P.Cast()((reds > 0), mstype.float32) * red_biggest * dhg3)
        dhg4 = green_smallest * (reds - blues) * crcp(P.Square()(blues - greens))
        dh_dg_4 = 60 * (P.Cast()((blues > 0), mstype.float32) * blue_biggest * dhg4)
        dhg5 = red_smallest * -1 * crcp(blues - reds)
        dh_dg_5 = 60 * (P.Cast()((blues > 0), mstype.float32) * blue_biggest * dhg5)

        dh_dg = (dh_dg_1 + dh_dg_2 + dh_dg_3 + dh_dg_4 + dh_dg_5) / 360

        dhb1 = (reds - greens) * crcp(P.Square()(saturation)) * crcp(P.Square()(value))
        dh_db_1 = 60 * (P.Cast()((blues > 0), mstype.float32) * blue_biggest * -1 * dhb1)
        dhb2 = blue_smallest * (greens - reds) * crcp(P.Square()(reds - blues))
        dh_db_2 = 60 * (P.Cast()((reds > 0), mstype.float32) * red_biggest * dhb2)
        dhb3 = green_smallest * -1 * crcp(reds - greens)
        dh_db_3 = 60 * (P.Cast()((reds > 0), mstype.float32) * red_biggest * dhb3)
        dhb4 = blue_smallest * (greens - reds) * crcp(P.Square()(greens - blues))
        dh_db_4 = 60 * (P.Cast()((greens > 0), mstype.float32) * green_biggest * dhb4)
        dhb5 = red_smallest * crcp(greens - reds)
        return function7_rgbtohsv(greens, green_biggest, dhb5, dh_db_1, dh_db_2, dh_db_3,\
                         dh_db_4, dout, dv_dr, dv_dg, dv_db, ds_dr, ds_dg, ds_db, dh_dr, dh_dg)
    return bprop


@bprop_getters.register(ScaleAndTranslate)
def get_bprop_scale_and_translate(self):
    """Grad definition for `ScaleAndTranslate` operation"""
    scale_and_translate_grad = G.ScaleAndTranslateGrad(self.kernel_type, self.antialias)

    def bprop(images, size, scale, translation, out, dout):
        images_fp32 = F.cast(images, mstype.float32)
        grad0_fp32 = scale_and_translate_grad(dout, images_fp32, scale, translation)
        grad0 = F.cast(grad0_fp32, F.dtype(images))
        result = (grad0, F.zeros_like(size), F.zeros_like(scale), F.zeros_like(translation))
        return result

    return bprop

File: mindspore/ccsrc/pipeline/pynative/grad/bprop_expander/grad_ops/grad_image_ops.cc
/**
 * Copyright 2022 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <set>
#include "pipeline/pynative/grad/bprop_expander/bprop_irbuilder.h"
#include "pipeline/pynative/grad/bprop_expander/grad_ops/common_utils.h"
#include "include/common/utils/utils.h"
#include "utils/ms_context.h"

namespace mindspore::expander::bprop {
REG_BPROP_BUILDERS_BEGIN(GradImageOps)
REG_BPROP_BUILDER("ResizeBicubic").SetUnusedInputs({i1, i2}).SetBody(BODYFUNC(ib) {
  auto images = ib->GetInput(kIndex0);
  auto size = ib->GetInput(kIndex1);
  auto dout = ib->GetInput(kIndex3);
  auto dx = ib->Emit(
    "ResizeBicubicGrad", {dout, images},
    {{"align_corners", ib->GetAttr("align_corners")}, {"half_pixel_centers", ib->GetAttr("half_pixel_centers")}});
  return {dx, ib->ZerosLike(size)};
});

REG_BPROP_BUILDER("CropAndResize").SetUnusedInputs({i3, i4}).SetBody(BODYFUNC(ib) {
  std::set<TypeId> allowed_types = {kNumberTypeFloat16, kNumberTypeFloat32, kNumberTypeFloat64};
  auto method = GetValue<std::string>(ib->GetAttr("method"));
  auto target = ib->GetTargetFromContext();
  auto is_ascend_cpu = (target == kAscendDevice || target == kCPUDevice);
  auto x = ib->GetInput(kIndex0);
  auto boxes = ib->GetInput(kIndex1);
  auto box_index = ib->GetInput(kIndex2);
  auto crop_size = ib->GetInput(kIndex3);
  auto dout = ib->GetInput(kIndex5);
  if (method != "bilinear") {
    if (!is_ascend_cpu) {
      return {ib->ZerosLike(x), ib->ZerosLike(boxes), ib->ZerosLike(box_index), ib->ZerosLike(crop_size)};
    }
  }
  auto image_type = ib->GetDtype(x);
  if (allowed_types.count(image_type->type_id()) != 0) {
    x = ib->Cast(x, kFloat32);
  }
  auto image_size = ib->Tensor(ib->GetShape(x), kInt32);
  const int64_t max_byte = 2e9;  // max bytes of image gradient
  auto dimage = ib->Emit("CropAndResizeGradImage", {dout, boxes, box_index, image_size},
                         {{"method", MakeValue(method)}, {"T", image_type}, {"max_Byte", MakeValue(max_byte)}});
  auto dbox = ib->Emit("CropAndResizeGradBoxes", {dout, x, boxes, box_index}, {{"method", MakeValue("bilinear")}});
  return {dimage, dbox, ib->ZerosLike(box_index), ib->ZerosLike(crop_size)};
});

REG_BPROP_BUILDER("ScaleAndTranslate").SetUnusedInputs({i1, i4}).SetBody(BODYFUNC(ib) {
  auto images = ib->GetInput(kIndex0);
  auto size = ib->GetInput(kIndex1);
  auto scale = ib->GetInput(kIndex2);
  auto translation = ib->GetInput(kIndex3);
  auto dout = ib->GetInput(kIndex5);
  auto images_dtype = ib->GetDtype(images);
  auto images_fp32 = (images_dtype->type_id() != kNumberTypeFloat32) ? ib->Cast(images, kFloat32) : images;
  auto grad0_fp32 = ib->Emit("ScaleAndTranslateGrad", {dout, images_fp32, scale, translation},
                             {{"kernel_type", ib->GetAttr("kernel_type")}, {"antialias", ib->GetAttr("antialias")}});
  auto grad0 = (images_dtype->type_id() != kNumberTypeFloat32) ? ib->Cast(grad0_fp32, images_dtype) : grad0_fp32;
  return {grad0, ib->ZerosLike(size), ib->ZerosLike(scale), ib->ZerosLike(translation)};
});

REG_BPROP_BUILDER("RGBToHSV").SetBody(BODYFUNC(ib) {
  auto images = ib->GetInput(kIndex0);
  auto out = ib->GetInput(kIndex1);
  auto dout = ib->GetInput(kIndex2);
  auto images_dtype = ib->GetDtype(images);
  if (images_dtype->type_id() != kNumberTypeFloat32) {
    images = ib->Cast(images, kFloat32);
    out = ib->Cast(out, kFloat32);
    dout = ib->Cast(dout, kFloat32);
  }
  auto tensor_0 = ib->Tensor(0, kFloat32);
  auto tensor_1 = ib->Tensor(1, kFloat32);
  auto tensor_n1 = ib->Tensor(-1, kFloat32);
  auto tensor_60 = ib->Tensor(60, kFloat32);
  auto tensor_360 = ib->Tensor(360, kFloat32);
  auto crcp = [&tensor_1, &ib](const NodePtr &x) { return ib->DivNoNan(tensor_1, x); };

  auto reds = ib->StridedSlice(images, {{-1, {0}}});
  auto greens = ib->StridedSlice(images, {{-1, {1}}});
  auto blues = ib->StridedSlice(images, {{-1, {2}}});
  auto dsr1 = ib->Greater(reds, tensor_0, kFloat32);
  auto saturation = ib->StridedSlice(out, {{-1, {1}}});
  auto value = ib->StridedSlice(out, {{-1, {2}}});
  auto red_biggest = ib->Cast(ib->LogicalAnd(ib->GreaterEqual(reds, blues), ib->GreaterEqual(reds, greens)), kFloat32);
  auto green_biggest = ib->Cast(ib->LogicalAnd(ib->Greater(greens, reds), ib->GreaterEqual(greens, blues)), kFloat32);
  auto blue_biggest = ib->Cast(ib->LogicalAnd(ib->Greater(blues, reds), ib->Greater(blues, greens)), kFloat32);
  auto red_smallest = ib->Cast(ib->LogicalAnd(ib->Less(reds, blues), ib->Less(reds, greens)), kFloat32);
  auto green_smallest = ib->Cast(ib->LogicalAnd(ib->LessEqual(greens, reds), ib->Less(greens, blues)), kFloat32);
  auto blue_smallest = ib->Cast(ib->LogicalAnd(ib->LessEqual(blues, reds), ib->LessEqual(blues, greens)), kFloat32);
  auto dv_dr = red_biggest;
  auto dv_dg = green_biggest;
  auto dv_db = blue_biggest;

  auto dsr2 = red_biggest * (green_smallest * greens + blue_smallest * blues) * crcp(ib->Square(reds));
  auto dsr3 = red_smallest * tensor_n1 * crcp((green_biggest * greens) + (blue_biggest * blues));
  auto ds_dr = dsr1 * (dsr2 + dsr3);
  auto dsg1 = ib->Greater(greens, tensor_0, kFloat32);
  auto dsg2 = green_biggest * (red_smallest * reds + blue_smallest * blues) * crcp(ib->Square(greens));
  auto dsg3 = green_smallest * tensor_n1 * crcp((red_biggest * reds) + (blue_biggest * blues));
  auto ds_dg = dsg1 * (dsg2 + dsg3);

  auto dsb1 = ib->Greater(blues, tensor_0, kFloat32);
  auto dsb2 = blue_biggest * (green_smallest * greens + red_smallest * reds) * crcp(ib->Square(blues));
  auto dsb3 = blue_smallest * tensor_n1 * crcp((green_biggest * greens) + (red_biggest * reds));
  auto ds_db = dsb1 * (dsb2 + dsb3);

  auto dhr1 = (greens - blues) * crcp(ib->Square(saturation)) * crcp(ib->Square(value));
  auto dh_dr_1 = tensor_60 * (ib->Greater(reds, tensor_0, kFloat32) * red_biggest * tensor_n1 * dhr1);
  auto dhr2 = red_smallest * (blues - greens) * crcp(ib->Square(reds - greens));
  auto dh_dr_2 = tensor_60 * (ib->Greater(greens, tensor_0, kFloat32) * green_biggest * dhr2);
  auto dhr3 = (blue_smallest * tensor_n1) * crcp(greens - blues);
  auto dh_dr_3 = tensor_60 * (ib->Greater(greens, tensor_0, kFloat32) * green_biggest * dhr3);
  auto dhr4 = red_smallest * (blues - greens) * crcp(ib->Square(blues - reds));
  auto dh_dr_4 = tensor_60 * (ib->Greater(blues, tensor_0, kFloat32) * blue_biggest * dhr4);
  auto dhr5 = green_smallest * crcp(blues - greens);
  auto dh_dr_5 = tensor_60 * (ib->Greater(blues, tensor_0, kFloat32) * blue_biggest * dhr5);
  auto dh_dr = (dh_dr_1 + dh_dr_2 + dh_dr_3 + dh_dr_4 + dh_dr_5) / tensor_360;

  auto dhg1 = (blues - reds) * crcp(ib->Square(saturation)) * crcp(ib->Square(value));
  auto dh_dg_1 = tensor_60 * (ib->Greater(greens, tensor_0, kFloat32) * green_biggest * tensor_n1 * dhg1);
  auto dhg2 = green_smallest * (reds - blues) * crcp(ib->Square(reds - greens));
  auto dh_dg_2 = tensor_60 * (ib->Greater(reds, tensor_0, kFloat32) * red_biggest * dhg2);
  auto dhg3 = blue_smallest * crcp(reds - blues);
  auto dh_dg_3 = tensor_60 * (ib->Greater(reds, tensor_0, kFloat32) * red_biggest * dhg3);
  auto dhg4 = green_smallest * (reds - blues) * crcp(ib->Square(blues - greens));
  auto dh_dg_4 = tensor_60 * (ib->Greater(blues, tensor_0, kFloat32) * blue_biggest * dhg4);
  auto dhg5 = red_smallest * tensor_n1 * crcp(blues - reds);
  auto dh_dg_5 = tensor_60 * (ib->Greater(blues, tensor_0, kFloat32) * blue_biggest * dhg5);
  auto dh_dg = (dh_dg_1 + dh_dg_2 + dh_dg_3 + dh_dg_4 + dh_dg_5) / tensor_360;

  auto dhb1 = (reds - greens) * crcp(ib->Square(saturation)) * crcp(ib->Square(value));
  auto dh_db_1 = tensor_60 * (ib->Greater(blues, tensor_0, kFloat32) * blue_biggest * tensor_n1 * dhb1);
  auto dhb2 = blue_smallest * (greens - reds) * crcp(ib->Square(reds - blues));
  auto dh_db_2 = tensor_60 * (ib->Greater(reds, tensor_0, kFloat32) * red_biggest * dhb2);
  auto dhb3 = green_smallest * tensor_n1 * crcp(reds - greens);
  auto dh_db_3 = tensor_60 * (ib->Greater(reds, tensor_0, kFloat32) * red_biggest * dhb3);
  auto dhb4 = blue_smallest * (greens - reds) * crcp(ib->Square(greens - blues));
  auto dh_db_4 = tensor_60 * (ib->Greater(greens, tensor_0, kFloat32) * green_biggest * dhb4);
  auto dhb5 = red_smallest * crcp(greens - reds);
  auto dh_db_5 = tensor_60 * (ib->Greater(greens, tensor_0, kFloat32) * green_biggest * dhb5);
  auto dh_db = (dh_db_1 + dh_db_2 + dh_db_3 + dh_db_4 + dh_db_5) / tensor_360;

  auto dout_r = ib->StridedSlice(dout, {{-1, {0}}});
  auto dout_g = ib->StridedSlice(dout, {{-1, {1}}});
  auto dout_b = ib->StridedSlice(dout, {{-1, {2}}});
  auto axis = MakeValue<int64_t>(-1);
  auto dv_drgb = ib->Emit("Stack", {ib->MakeTuple({dout_b * dv_dr, dout_b * dv_dg, dout_b * dv_db})}, {{"axis", axis}});
  auto ds_drgb = ib->Emit("Stack", {ib->MakeTuple({dout_g * ds_dr, dout_g * ds_dg, dout_g * ds_db})}, {{"axis", axis}});
  auto dh_drgb = ib->Emit("Stack", {ib->MakeTuple({dout_r * dh_dr, dout_r * dh_dg, dout_r * dh_db})}, {{"axis", axis}});
  auto doutient_input = dv_drgb + ds_drgb + dh_drgb;
  return {doutient_input};
});
REG_BPROP_BUILDERS_END
}  // namespace mindspore::expander::bprop
